include/nighthawk/common/factories.h

```diff
@@ -48,4 +49,26 @@ class TerminationPredicateFactory {
          const Envoy::MonotonicTime scheduled_starting_time) const PURE;
 };
 
+/**
+ * Factory Interface to create Envoy::Stats::Sink in Nighthawk.
+ * Implemented for each Envoy::Stats::Sink and registered via
+ * Registry::registerFactory() or the convenience class RegisterFactory.
+ */
+class NighthawkStatsSinkFactory : public Envoy::Config::TypedFactory {
+public:
+  ~NighthawkStatsSinkFactory() override = default;
+
+  /**
+   * Create a particular Envoy::Stats::Sink implementation. If the
+   * implementation is unable to produce a Sink with the provided parameters, it
+   * should throw an EnvoyException. The returned pointer should always be
+   * valid.
+   * @param symbol_table supplies the symbol_table instance
+   */
+  virtual std::unique_ptr<Envoy::Stats::Sink>
+  createStatsSink(Envoy::Stats::SymbolTable& symbol_table) PURE;
+
+  std::string category() const override { return "nighthawk.stats_sinks"; }
```

Comment:
As this is in the interface, can we make this `PURE` and punt implementing this to a base implementation or the `FlushWorker` directly?

Reply:
Here I follow the Envoy example https://github.com/envoyproxy/envoy/blob/2a6a4b8d758b4bf4238cb6d2babdd3d83fbf45bc/source/server/configuration_impl.h#L47

I think the category() should be here since different implementation of NighthawkStatsSinkFactory should share the same category.

Reply:
I see; I think the difference between the Envoy code you point out and this file, is that the Envoy code is an implementation (`configuration_impl.h` in `source/...`), whereas NHs interfaces generally reside in `includes/`. 
Hence the suggestion to declare this PURE here, and create a base class to implement the common factor that should be shared across implementations under `source/..` (implementing only the getter for the `category()`).  That way, the interface is clean (which helps with that mocking the interface stays trivial), and the implementation can still be shared, which is what you're after, right?

Reply:
ah, the `StatsSinkFactory` in the link above is an abstract class not an implementation.

Anyway, agree that it is better to make the category() PURE here.

Reply:
Fair enough